### PR TITLE
[MUSIC][JSON-RPC] AudioLibrary.RefreshArtist and AudioLibrary.RefreshAlbum

### PR DESCRIPTION
**Pull request by acidzab**
## Description

This PR gives the possibility to refresh artist and album informations through JSON-RPC

- added AudioLibrary.RefreshArtist
- added AudioLibrary.RefreshAlbum
- bumped api version and updated the json schema

Off-topic: this is my first pr and i'm not a c++ expert, just trying to give my contribute, so any constructive criticism and hints are accepted :)

## Motivation and context

Give the possibility to refresh Artist and Album informations from third party tools

## How has this been tested?

Local curls on my build, artist with an AIF and NFO, album with the Generic Scraper

```shell
curl --data-binary '{"jsonrpc":"2.0","method":"AudioLibrary.RefreshArtist","id":"1","params":{"artistid":2}}' -H 'content-type: application/json;' http://user:pass@127.0.0.1:8088/jsonrpc

curl --data-binary '{"jsonrpc":"2.0","method":"AudioLibrary.RefreshAlbum","id":"1","params":{"albumid":2}}' -H 'content-type: application/json;' http://user:pass@127.0.0.1:8088/jsonrpc

curl --data-binary '[{"jsonrpc":"2.0","method":"AudioLibrary.RefreshArtist","id":"1","params":{"artistid":2}},{"jsonrpc":"2.0","method":"AudioLibrary.RefreshArtist","id":"1","params":{"artistid":3}}]' -H 'content-type: application/json;' http://user:pass@127.0.0.1:8088/jsonrpc
```

## What is the effect on users?

These apis acts like a regular information refresh from the GUI (without the confirmation dialog 😅 )

## Types of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Clean up (non-breaking change which removes non-working, unmaintained functionality)
- [ ] Improvement (non-breaking change which improves existing functionality)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that will cause existing functionality to change)
- [ ] Cosmetic change (non-breaking change that doesn't touch code)
- [ ] Student submission (PR was done for educational purposes and will be treated as such)
- [ ] None of the above (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [x] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by the-black-eagle** [comment]
Thanks!  Please apply the Jenkins diff to fix up the formatting.  Build failures are not related to this PR.

**Comment by acidzab** [comment]
@the-black-eagle done, many thanks :)

**Comment by the-black-eagle** [comment]
>  One question: do you think if these two apis could be backported to Omega ?
No, can't backport new features, only fixes.

**Comment by acidzab** [comment]
proposed the same formatting changes i've applied before and on my repo there isn't any difference between this new diff and the current branch

**Comment by the-black-eagle** [comment]
|proposed the same formatting changes i've applied before and on my repo there isn't any difference between this new diff and the current branch

Yeah, Jenkins looks at each commit in turn, so it will flag up the diff, even though the last commit fixes up the formatting.  I just squash it into the first commit, usually as a fixup.

**Comment by acidzab** [comment]
> |proposed the same formatting changes i've applied before and on my repo there isn't any difference between this new diff and the current branch
> 
> Yeah, Jenkins looks at each commit in turn, so it will flag up the diff, even though the last commit fixes up the formatting. I just squash it into the first commit, usually as a fixup.

ok i think i've got it 😅

**Review by the-black-eagle — APPROVED** [review]
LGTM.

**Comment by the-black-eagle** [comment]
Build failures are unrelated to this PR.  @DaveTBlake , I'll give it a couple of days in case you pop out of the woodwork with some comments, but given no objections, I'll merge this at the weekend.

@acidzab Thanks for your contribution, it's much appreciated.

**Comment by the-black-eagle** [comment]
jenkins build this please

**Comment by the-black-eagle** [comment]
Jenkins build this please

File: xbmc/interfaces/json-rpc/AudioLibrary.cpp [hunk 1]
```diff
@@ -1354,6 +1356,60 @@ JSONRPC_STATUS CAudioLibrary::GetAdditionalSongDetails(const CVariant& parameter
...
+    return InternalError;
+
+  //checking if artistID is a valid one
+  if (!musicdatabase.GetArtistExists(artistID))
+    return InvalidParams;
+
+  //set the artist id on the musicdb url
+  musicUrl.AddOption("artistid", artistID);
+
+  //executing the StartArtistScan for refreshing the artist scraped informations
```
Comment (DaveTBlake) [inline xbmc/interfaces/json-rpc/AudioLibrary.cpp:1382]: //Start rescraping additional information for the given artist

File: xbmc/interfaces/json-rpc/AudioLibrary.cpp [hunk 2]
```diff
@@ -1354,6 +1356,60 @@ JSONRPC_STATUS CAudioLibrary::GetAdditionalSongDetails(const CVariant& parameter
...
+  CMusicDatabase musicdatabase;
+  if (!musicdatabase.Open())
+    return InternalError;
+
+  //check if albumID is a valid one
+  CAlbum album;
+  if (!musicdatabase.GetAlbum(albumID, album, false))
+    return InvalidParams;
+
+  std::string path = StringUtils::Format("musicdb://albums/{}/", albumID);
```
Comment (DaveTBlake) [inline xbmc/interfaces/json-rpc/AudioLibrary.cpp:1405]: Can this be done using  'musicUrl.AddOption' consistant with how you have done in RefreshArtist? I'm asking if possible not asserting it is done that way, I really can't remember the syntax details.

File: xbmc/interfaces/json-rpc/schema/methods.json [hunk 3]
```diff
@@ -3063,6 +3063,34 @@
     ],
     "returns": "string"
   },
+  "AudioLibrary.RefreshArtist": {
+    "type": "method",
+    "description": "Refresh the given artist in the library",
```
Comment (DaveTBlake) [inline xbmc/interfaces/json-rpc/schema/methods.json:3068]: "description": "Refresh the additional information for the given artist",

File: xbmc/interfaces/json-rpc/schema/methods.json [hunk 4]
```diff
@@ -3063,6 +3063,34 @@
...
+        "name": "artistid",
+        "$ref": "Library.Id",
+        "required": true
+      }
+    ],
+    "returns": "string"
+  },
+  "AudioLibrary.RefreshAlbum": {
+    "type": "method",
+    "description": "Refresh the given album in the library",
```
Comment (DaveTBlake) [inline xbmc/interfaces/json-rpc/schema/methods.json:3082]: "description": "Refresh the additional information for the given album",

File: xbmc/interfaces/json-rpc/AudioLibrary.cpp [hunk 5]
```diff
@@ -1354,6 +1356,60 @@ JSONRPC_STATUS CAudioLibrary::GetAdditionalSongDetails(const CVariant& parameter
...
+
+  //execute the album refresh job
```
Comment (DaveTBlake) [inline xbmc/interfaces/json-rpc/AudioLibrary.cpp:1407]: //Start rescraping additional information for the given album

**Review by DaveTBlake — CHANGES_REQUESTED** [review]
Because of the general confusion over the difference between music library (built from scanning embedded tag metadata, and then optionally fetching additional info for artists and albums using a scraper)  and the video libray (built by scraping external sources using folder names) I would like to be a little bit fussy over descriptions here. You are only refreshing the additional (scraped) information, not updating the entire artist or album.

I did wonder about calling the methods RefreshArtistInfo and RefreshAlbumInfo for clarity, but maybe that is too clunky. 

I'm also very out of touch, but at one time there was a requirement to keep the LibraryBuiltins  in line with the functionality offered by JSON. The other library functions cleaning, exporting, and updating (starting a full scan of tag data) are all in builtins, so should these refresh facilities also be there? @the-black-eagle  maybe you have thoughts on this, I'm not concerned either way but just bringing up some historic knowledge.

**Comment by DaveTBlake** [comment]
@the-black-eagle darn you merged while I was reviewing, how about that for bad luck!

**Comment by the-black-eagle** [comment]
> Because of the general confusion over the difference between music library (built from scanning embedded tag metadata, and then optionally fetching additional info for artists and albums using a scraper) and the video libray (built by scraping external sources using folder names) I would like to be a little bit fussy over descriptions here. You are only refreshing the additional (scraped) information, not updating the entire artist or album.
> 
> I did wonder about calling the methods RefreshArtistInfo and RefreshAlbumInfo for clarity, but maybe that is too clunky.

Yeah, I did consider that but both methods mimic what the GUI displays (which is just a refresh button for a given artist or album).  I think for new coders, it's easier to relate the two (GUI & json) like this.  Although I understand your POV, I don't think at the moment it matters too much.  Coming from a "using the GUI" POV, RefreshArtist makes perfect sense.

> darn you merged while I was reviewing, how about that for bad luck!

Well, pick up the ball sooner then my friend!!

**Comment by acidzab** [comment]
Hi, sorry i could reply just now but i've read @DaveTBlake  and @the-black-eagle replies to this pr.
Thanks Dave for having reviewed my PR, if it's not a problem i could open another PR for applying the changes proposed by Dave. Let me know :)

**Comment by garbear** [comment]
> Well, pick up the ball sooner then my friend!!  

Given that we're all volunteers with lives, a few days is good etiquette for random areas of the code, but we want at least one or two weeks if a member is tagged as a code owner (as per their definition). This PR was merged prematurely, but follow-ups make this no-problem.

**Comment by acidzab** [comment]
@DaveTBlake i've implemented the requested changes here #25698 :)

**Comment by acidzab** [comment]
Hi @DaveTBlake @the-black-eagle i'll resolve reviews conversations, the updated code is available in the pr #25698 :)

**Pull request merged**